pkg/apis/provisioning/v1alpha5/requirements.go

```diff
@@ -76,246 +76,232 @@ var (
-	}
-	allKeys := stringsets.NewString(r.Keys()...).Union(stringsets.NewString(requirements.Keys()...))
-	for key := range allKeys {
-		if err := r.isKeyDefined(key, requirements); err != nil {
-			errs = errs.Also(err)
+// Compatible ensures the provided requirements can be met. It is
+// non-commutative (i.e., A.Compatible(B) != B.Compatible(A))
+func (r Requirements) Compatible(requirements Requirements) (errs *apis.FieldError) {
+	for _, key := range r.Keys().Union(requirements.Keys()).UnsortedList() {
+		// Key must be defined if required
+		if values := requirements.Values(key); values.Len() != 0 && !values.Complement && !r.hasRequirement(withKey(key)) {
+			errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("undefined not in %s", values), "values"))
 		}
-		if err := r.hasCommons(key, requirements); err != nil {
-			errs = errs.Also(err)
+		// Values must overlap
+		if values := r.Values(key); values.Intersection(requirements.Values(key)).Len() == 0 {
+			errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("%s not in %s", values, requirements.Values(key)), "values"))
 		}
-		if err := r.satisfyExistOperators(key, requirements); err != nil {
-			errs = errs.Also(err)
+		// // Exist is not compatible with NotExists
+		// if r.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpExists)) && requirements.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpDoesNotExist)) {
+		// 	errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("operator %s incompatible with %s", v1.NodeSelectorOpDoesNotExist, v1.NodeSelectorOpExists), "operator"))
+		// }
+		// // NotExists is not compatible with Exists or In
+		// if r.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpDoesNotExist)) {
+		// 	if requirements.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpExists)) {
+		// 		errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("operator %s incompatible with %s", v1.NodeSelectorOpExists, v1.NodeSelectorOpDoesNotExist), "operator"))
+		// 	}
+		// 	if requirements.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpIn)) {
+		// 		errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("operator %s incompatible with %s", v1.NodeSelectorOpIn, v1.NodeSelectorOpDoesNotExist), "operator"))
+		// 	}
+		// }
+		// // In or Exists is required to be compatible with Exists
+		// if !r.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpIn)) || !r.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpExists)) {
+		// 	if requirements.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpExists)) {
+		// 		errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("operator %s requires %s or %s", v1.NodeSelectorOpExists, v1.NodeSelectorOpIn, v1.NodeSelectorOpExists), "operator"))
+		// 	}
+		// }
+
+		// Exists requires Exists or In
+		if requirements.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpExists)) {
+			if !r.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpExists)) || !r.hasRequirement(withKeyAndOperator(key, v1.NodeSelectorOpIn)) {
```

Comment:
So we are assuming that Exists only pairs with Exists or In? This may be ok when checking compatibility between provisioner and pods. But later we will also use this for grouping pods.
Can't we use the original definition of Exists that only requires r.hasRequirement(withKey(key)) == true?

Also I think line 259 is not correct. It should be && not ||.